src/boot/efi/stub.c

```diff
@@ -345,36 +377,59 @@ static EFI_STATUS cmdline_append_and_measure_addons(
                         continue;
                 }
 
-                _cleanup_free_ char16_t *tmp = TAKE_PTR(buffer),
-                                        *extra16 = xstrn8_to_16((char *)loaded_addon->ImageBase + addrs[UNIFIED_SECTION_CMDLINE],
-                                                                szs[UNIFIED_SECTION_CMDLINE]);
-                buffer = xasprintf("%ls%ls%ls", strempty(tmp), isempty(tmp) ? u"" : u" ", extra16);
-        }
-
-        mangle_stub_cmdline(buffer);
+                if (ret_cmdline && szs[UNIFIED_SECTION_CMDLINE] > 0) {
+                        _cleanup_free_ char16_t *tmp = TAKE_PTR(cmdline),
+                                                *extra16 = xstrn8_to_16((char *)loaded_addon->ImageBase + addrs[UNIFIED_SECTION_CMDLINE],
+                                                                        szs[UNIFIED_SECTION_CMDLINE]);
+                        cmdline = xasprintf("%ls%ls%ls", strempty(tmp), isempty(tmp) ? u"" : u" ", extra16);
+                }
 
-        if (!isempty(buffer)) {
-                _cleanup_free_ char16_t *tmp = TAKE_PTR(*cmdline_append);
-                bool m = false;
+                if (ret_dt_bases && szs[UNIFIED_SECTION_DTB] > 0) {
+                        dt_sizes = xrealloc(dt_sizes,
+                                            n_dt * sizeof(size_t),
+                                            (n_dt + 1)  * sizeof(size_t));
+                        dt_sizes[n_dt] = szs[UNIFIED_SECTION_DTB];
+
+                        dt_bases = xrealloc(dt_bases,
+                                            n_dt * sizeof(EFI_PHYSICAL_ADDRESS),
+                                            (n_dt + 1) * sizeof(EFI_PHYSICAL_ADDRESS));
+                        dt_bases[n_dt] = xmalloc(dt_sizes[n_dt]);
+                        memcpy(dt_bases[n_dt],
+                               PHYSICAL_ADDRESS_TO_POINTER(POINTER_TO_PHYSICAL_ADDRESS(loaded_addon->ImageBase) + addrs[UNIFIED_SECTION_DTB]),
+                               dt_sizes[n_dt]);
+
+                        ++n_dt;
+                }
+        }
 
-                (void) tpm_log_load_options(buffer, &m);
-                *ret_parameters_measured = m;
+        if (ret_cmdline && !isempty(cmdline))
+                *ret_cmdline = TAKE_PTR(cmdline);
 
-                *cmdline_append = xasprintf("%ls%ls%ls", strempty(tmp), isempty(tmp) ? u"" : u" ", buffer);
+        if (ret_n_dt && n_dt > 0) {
+                *ret_dt_bases = TAKE_PTR(dt_bases);
+                *ret_dt_sizes = TAKE_PTR(dt_sizes);
+                *ret_n_dt = n_dt;
         }
 
         return EFI_SUCCESS;
+
+fail:
+        for (size_t i = 0; i < n_dt; ++i)
+                free(dt_bases[i]);
```

Comment:
hmm, use CLEANUP_ARRAY()? (and add it to uefi headers if missing there?)

Reply:
I tried to move it but I'm getting an error I don't undestand:

```
../src/boot/efi/stub.c:344:9: error: cleanup argument not a function
  344 |         CLEANUP_ARRAY(dt_bases, n_dt, dt_array_free);
```

I have no idea why it says it's not a function, when it clearly is, with prototype `static void dt_array_free(void **dt_bases, size_t n_dts) `

Reply:
did you figure that out ever?

Reply:
No I couldn't make it work, after bashing it with a hammer for an hour or so I gave up. If anybody knows how to make it work I'm happy to take the change, but have no idea how otherwise

Reply:
did you actually copy array_cleanup() static inline func over into fundamental/ too?

Reply:
(i mean, `array_cleanup()` is the function used as cleanup argument, not the one you pass

Reply:
Seems to just work here: 

https://paste.centos.org/view/raw/dbff36e8

Reply:
(tested that with both gcc and clang, works perfectly on both)

Reply:
interesting - I no longer have saved what I was doing so no idea if it is different and how, I'll apply that when I get to it and try again, thanks